File: backend/internal/domain/daily.go
package domain

import (
	"context"
	"time"

	"github.com/google/uuid"
)

type WordOfTheDay struct {
	ID        uuid.UUID
	WordID    int
	Date      time.Time
	CreatedAt time.Time
	Word      *Word
}

type DailyRepository interface {
	// Word of the day
	GetWordOfTheDay(ctx context.Context, date time.Time) (*WordOfTheDay, error)
	SetWordOfTheDay(ctx context.Context, wordID int, date time.Time) error
	GetRecentWords(ctx context.Context, limit int) ([]string, error)
	GetWordByWord(ctx context.Context, word string) (*Word, error)
	WordExistsInUserCards(ctx context.Context, userID uuid.UUID, word string) (bool, error)

	// Daily mix
	GetDailyMixCandidates(ctx context.Context, userID uuid.UUID) ([]DailyMixCandidate, error)
	GetDailyProgressCount(ctx context.Context, userID uuid.UUID, cardIDs []int) (int, error)
}

type DailyService interface {
	GetWordOfTheDay(ctx context.Context, userID uuid.UUID) (*WordOfTheDayResponse, error)
	AddWordOfTheDayToDecks(ctx context.Context, userID uuid.UUID, deckID int) (*Card, error)
	GetDailyMix(ctx context.Context, userID uuid.UUID) (*DailyMixResponse, error)
}

type SuggestedDeck struct {
	ID   int    `json:"id"`
	Name string `json:"name"`
}

type WordOfTheDayResponse struct {
	Word          string         `json:"word"`
	Transcription string         `json:"transcription"`
	PartOfSpeech  string         `json:"part_of_speech"`
	Definition    string         `json:"definition"`
	SuggestedDeck *SuggestedDeck `json:"suggested_deck"`
	AlreadyAdded  bool           `json:"already_added"`
}

type DailyMixCandidate struct {
	Card     Card
	Priority int
}

type DailyMixProgress struct {
	Done  int `json:"done"`
	Total int `json:"total"`
}

type DailyMixResponse struct {
	Cards    []Card           `json:"cards"`
	Progress DailyMixProgress `json:"progress"`
}

File: frontend/src/lib/api/daily.ts
import { fetchWithAuth } from './fetchWithAuth';
import { mapCard } from './backend';
import type { VocabularyCard } from '@/types/card';

export interface WordOfTheDayData {
  word: string;
  translation: string;
  transcription: string;
  part_of_speech: string;
  definition: string;
  suggested_deck: { id: number; name: string } | null;
  already_added: boolean;
}

export interface DailyMixData {
  cards: VocabularyCard[];
  progress: { done: number; total: number };
}

export async function fetchWordOfTheDay(): Promise<WordOfTheDayData> {
  const res = await fetchWithAuth('/api/daily/word-of-the-day');
  if (!res.ok) throw new Error(`Word of the day error: ${res.status}`);
  return res.json();
}

export async function addWordOfTheDayToDeck(deckId: number): Promise<VocabularyCard> {
  const res = await fetchWithAuth('/api/daily/word-of-the-day/add', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ deck_id: deckId }),
  });
  if (!res.ok) throw new Error(`Add word of the day error: ${res.status}`);
  const data = await res.json();
  return mapCard(data as Parameters<typeof mapCard>[0]);
}

export async function fetchDailyMix(): Promise<DailyMixData> {
  const res = await fetchWithAuth('/api/daily/mix');
  if (!res.ok) throw new Error(`Daily mix error: ${res.status}`);
  const data = await res.json();
  return {
    cards: (data.cards ?? []).map((c: Parameters<typeof mapCard>[0]) => mapCard(c)),
    progress: data.progress,
  };
}
